### PR TITLE
More structured way to handle parse exceptions

### PR DESCRIPTION
**Pull request by jihoonson**
### Description

This PR has refactoring around handling `ParseExceptions` in ingestion, so that we can easily support the same logic for parallel task as well. The parseException can be thrown in 2 places on the ingestion side, i.e., when you parse the input data based on input format (in `InputEntityReader`) and when you convert the parsed object to a desired data type (in `IncrementalIndex`). To handle the exceptions in a more structured way, I added `ParseExceptionHandler` which handles all parseExceptions thrown on the ingestion side. 

`FilteringCloseableInputRowIterator` was added to handle common row filters. It also has the parseExceptionHandler to handle parseExceptions thrown while parsing input data. The same parseExceptionHandler instance is passed down to `AppenderatorImpl` to handle parseExceptions thrown while converting parsed objects.

<hr>

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/licenses.yaml)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

<hr>

##### Key changed/added classes in this PR
 * `ParseExceptionHandler`
 * `FilteringCloseableInputRowIterator`

### CODE REVIEW

File: processing/src/main/java/org/apache/druid/segment/incremental/ParseExceptionHandler.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+ */
+public class ParseExceptionHandler
+{
+  private static final Logger LOG = new Logger(ParseExceptionHandler.class);
+
+  private final RowIngestionMeters rowIngestionMeters;
+  private final boolean logParseExceptions;
+  private final int maxAllowedParseExceptions;
+  @Nullable
+  private final CircularBuffer<Throwable> savedParseExceptions;
```
Comment (abhishekagarwal87) [inline processing/src/main/java/org/apache/druid/segment/incremental/ParseExceptionHandler.java:47]: just curious, can this be `private final CircularBuffer<Throwable>` itself?

Comment (jihoonson) [reply processing/src/main/java/org/apache/druid/segment/incremental/ParseExceptionHandler.java]: Oh yeah, it can. I just copied existing code without noticing it. Changed it now.

**Comment by jihoonson** [comment]
```278 L | B(1/2)        if (addResult.isRowAdded()) {
279 L                   rowIngestionMeters.incrementProcessed();
280 F | L | B(0/2)    } else if (addResult.hasParseException()) {
281 L                   parseExceptionHandler.handle(addResult.getParseException());
282 F                 }
```

The test coverage bot is failing because of the missing tests for these branches, but they are being covered by `IndexTaskTest` which is in `indexing-service`.

File: core/src/main/java/org/apache/druid/data/input/impl/MapInputRowParser.java [hunk 2]
```diff
@@ -93,23 +94,32 @@ public static InputRow parse(
...
-      }
     }
     catch (Exception e) {
-      throw new ParseException(e, "Unparseable timestamp found! Event: %s", theMap);
+      throw new ParseException(e, "Unparseable timestamp found! Event: %s", rawMapToPrint(theMap));
+    }
+    if (timestamp == null) {
+      throw new ParseException("Unparseable timestamp found! Event: %s", rawMapToPrint(theMap));
+    }
+    if (!Intervals.ETERNITY.contains(timestamp)) {
```
Comment (suneet-s) [inline core/src/main/java/org/apache/druid/data/input/impl/MapInputRowParser.java:104]: wow, I never thought of this case

File: core/src/main/java/org/apache/druid/data/input/impl/MapInputRowParser.java [hunk 3]
```diff
@@ -93,23 +94,32 @@ public static InputRow parse(
...
-            )
-        );
-      }
     }
     catch (Exception e) {
-      throw new ParseException(e, "Unparseable timestamp found! Event: %s", theMap);
+      throw new ParseException(e, "Unparseable timestamp found! Event: %s", rawMapToPrint(theMap));
+    }
+    if (timestamp == null) {
+      throw new ParseException("Unparseable timestamp found! Event: %s", rawMapToPrint(theMap));
```
Comment (suneet-s) [inline core/src/main/java/org/apache/druid/data/input/impl/MapInputRowParser.java:102]: Instead of printing the whole map, I think it would be better to just print the timestamp string that was unparsable.
Similar comment on line 99, this would get rid of the need of `rawMapToPrint` and guarantees that the invalid timestamp is always logged regardless of where in the event it is.

```suggestion
      throw new ParseException("Unparseable timestamp found! Timestamp column (%s): %s", timestampSpec.getTimestampColumn(), theMap.get(timestampSpec.getTimestampColumn()));
```

File: core/src/main/java/org/apache/druid/data/input/impl/MapInputRowParser.java [hunk 4]
```diff
@@ -93,23 +94,32 @@ public static InputRow parse(
...
+    }
+    if (!Intervals.ETERNITY.contains(timestamp)) {
+      throw new ParseException(
+          "Encountered row with timestamp that cannot be represented as a long: [%s]",
+          rawMapToPrint(theMap)
+      );
```
Comment (suneet-s) [inline core/src/main/java/org/apache/druid/data/input/impl/MapInputRowParser.java:108]: After your refactoring in this change, the `parse` functions on line 65 and 70 can be made private and package private (VisibleForTesting) respectively

File: extensions-core/kafka-indexing-service/src/test/java/org/apache/druid/indexing/kafka/KafkaIndexTaskTest.java [hunk 5]
```diff
@@ -106,6 +103,9 @@
 import org.apache.druid.query.timeseries.TimeseriesQueryEngine;
 import org.apache.druid.query.timeseries.TimeseriesQueryQueryToolChest;
 import org.apache.druid.query.timeseries.TimeseriesQueryRunnerFactory;
+import org.apache.druid.segment.incremental.RowIngestionMeters;
+import org.apache.druid.segment.incremental.RowIngestionMetersFactory;
+import org.apache.druid.segment.incremental.RowIngestionMetersTotals;
```
Comment (suneet-s) [inline extensions-core/kafka-indexing-service/src/test/java/org/apache/druid/indexing/kafka/KafkaIndexTaskTest.java:108]: nit: Maybe these 3 classes should be moved into their own sub package `org.apache.druid.segment.incremental.stats`

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/AbstractBatchIndexTask.java [hunk 6]
```diff
@@ -135,6 +148,49 @@ public void stopGracefully(TaskConfig taskConfig)
     }
   }
 
+  public static FilteringCloseableInputRowIterator inputSourceReader(
```
Comment (suneet-s) [inline indexing-service/src/main/java/org/apache/druid/indexing/common/task/AbstractBatchIndexTask.java:151]: javadocs please

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTask.java [hunk 7]
```diff
@@ -151,7 +151,7 @@ private static String makeTaskId(RealtimeAppenderatorIngestionSpec spec)
   private volatile Thread runThread = null;
 
   @JsonIgnore
-  private CircularBuffer<Throwable> savedParseExceptions;
+  private ParseExceptionHandler parseExceptionHandler;
```
Comment (suneet-s) [inline indexing-service/src/main/java/org/apache/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTask.java:154]: Should this also be annotated with `@MonotonicNonNull`

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/AbstractBatchIndexTask.java [hunk 8]
```diff
@@ -135,6 +148,49 @@ public void stopGracefully(TaskConfig taskConfig)
...
+    final List<String> metricsNames = Arrays.stream(dataSchema.getAggregators())
+                                            .map(AggregatorFactory::getName)
+                                            .collect(Collectors.toList());
+    final InputSourceReader inputSourceReader = dataSchema.getTransformSpec().decorate(
+        inputSource.reader(
+            new InputRowSchema(
+                dataSchema.getTimestampSpec(),
+                dataSchema.getDimensionsSpec(),
+                metricsNames
+            ),
```
Comment (suneet-s) [inline indexing-service/src/main/java/org/apache/druid/indexing/common/task/AbstractBatchIndexTask.java:170]: This is calculated in 3 places in the code - `AbstractBatchIndexTask`, `InputSourceSampler` and `SeekableStreamIndexTaskRunner`. I think it would be a good idea to consolidate these methods, so that metricNames is always calculated the same way given a dataSchema

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTask.java [hunk 9]
```diff
@@ -550,7 +546,9 @@ public Response getUnparseableEvents(
   )
   {
     IndexTaskUtils.datasourceAuthorizationCheck(req, Action.READ, getDataSource(), authorizerMapper);
-    List<String> events = IndexTaskUtils.getMessagesFromSavedParseExceptions(savedParseExceptions);
+    List<String> events = IndexTaskUtils.getMessagesFromSavedParseExceptions(
+        parseExceptionHandler.getSavedParseExceptions()
```
Comment (suneet-s) [inline indexing-service/src/main/java/org/apache/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTask.java:550]: parseExceptionHandler can be null if the `task.run()` hasn't been called. This is probably unlikely?

File: processing/src/main/java/org/apache/druid/segment/incremental/OffheapIncrementalIndex.java [hunk 10]
```diff
@@ -233,16 +232,13 @@ protected AddToFactsResult addToFacts(
         }
         catch (ParseException e) {
           // "aggregate" can throw ParseExceptions if a selector expects something but gets something else.
-          if (getReportParseExceptions()) {
-            throw new ParseException(e, "Encountered parse error for aggregator[%s]", getMetricAggs()[i].getName());
-          } else {
-            log.debug(e, "Encountered parse error, skipping aggregator[%s].", getMetricAggs()[i].getName());
-          }
+          log.debug(e, "Encountered parse error, skipping aggregator[%s].", getMetricAggs()[i].getName());
+          parseExceptionMessages.add(e.getMessage());
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/segment/incremental/OffheapIncrementalIndex.java:236]: This is a change in behavior. Previously the aggs would fail fast, but now it tries to parse all of them before bubbling up all the parse exceptions. Was this change intentional? I see a similar change in at least the Onheap implementation.

File: processing/src/main/java/org/apache/druid/segment/incremental/NoopRowIngestionMeters.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment.incremental;
+
+import java.util.Collections;
+import java.util.Map;
+
+public class NoopRowIngestionMeters implements RowIngestionMeters
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/segment/incremental/NoopRowIngestionMeters.java:25]: Javadocs please.

Should this just be a singleton?

File: core/src/main/java/org/apache/druid/data/input/impl/MapInputRowParser.java (hunk 3 above)
Comment (jihoonson) [reply core/src/main/java/org/apache/druid/data/input/impl/MapInputRowParser.java]: It is a good idea to print the timestamp string, but I would like to keep the current behavior as well (this logging is not what I added in this PR). I modified the error message to include timestamp.

File: core/src/main/java/org/apache/druid/data/input/impl/MapInputRowParser.java (hunk 4 above)
Comment (jihoonson) [reply core/src/main/java/org/apache/druid/data/input/impl/MapInputRowParser.java:120]: Done.

File: extensions-core/kafka-indexing-service/src/test/java/org/apache/druid/indexing/kafka/KafkaIndexTaskTest.java (hunk 5 above)
Comment (jihoonson) [reply extensions-core/kafka-indexing-service/src/test/java/org/apache/druid/indexing/kafka/KafkaIndexTaskTest.java:108]: That might be nice. I think we probably need new classes for metrics of native batch ingestion. I will reorganize the package in a follow-up PR if is good.

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/AbstractBatchIndexTask.java (hunk 6 above)
Comment (jihoonson) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/AbstractBatchIndexTask.java:158]: Added.

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTask.java (hunk 7 above)
Comment (jihoonson) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTask.java]: Added.

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/AbstractBatchIndexTask.java (hunk 8 above)
Comment (jihoonson) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/AbstractBatchIndexTask.java:177]: Yeah, it would be useful. Or we can change `InputRowSchema` to have all type information of metrics as well, so that we can check exceptions in parsing and type casting in only `InputEntityReader`. It means, after this change, you will not have to worry about parseExceptions in `IncrementalIndex` anymore. Probably we don't need `ParseExceptionHandler` at all in that case. I haven't done this refactoring here because it will be quite complicated and this PR is already big enough.

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTask.java (hunk 9 above)
Comment (jihoonson) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTask.java:556]: This is fine. Any APIs cannot be called until `chatHandler` is registered. The `chatHandler` is registered in `run()` after `parseExceptionHandler` is created.

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/BatchAppenderators.java [hunk 12]
```diff
@@ -59,7 +65,9 @@ public static Appenderator newAppenderator(
       TaskToolbox toolbox,
       DataSchema dataSchema,
       AppenderatorConfig appenderatorConfig,
-      DataSegmentPusher segmentPusher
+      DataSegmentPusher segmentPusher,
+      RowIngestionMeters rowIngestionMeters,
+      ParseExceptionHandler parseExceptionHandler
```
[the remark this answers is not in the document]
Comment (jihoonson) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/BatchAppenderators.java:70]: I see your point, but I'm not sure if it's good to have `parseExceptionHandler` be a `rowIngestionMeters` provider. I guess we have a couple of options for alternatives:

- Merge `RowIngestioMeters` and `ParseExceptionHandler`. `RowIngestionMeters` will have `handle(ParseException)` method and throw an exception when it hits the limit. This seems a bit strange to me since metrics registry can throw an exception.
- Make `RowIngestionMeters` to be only a registry and use individual `Meter`s. For example, `ParseExceptionHandler` will have only a `Meter` for counting unparseable events. `AppenderatorImpl` will have only a `Meter` for counting processed events. This seems sane, but can make code error-prone since we will have multiple `Meter`s which should be properly chosen whenever we use them.

I'm not sure any of these are better than now. Do you have any better idea?

File: server/src/main/java/org/apache/druid/segment/realtime/appenderator/PeonAppenderatorsManager.java [hunk 13]
```diff
@@ -129,7 +137,9 @@ public Appenderator createOfflineAppenderatorForTask(
           dataSegmentPusher,
           objectMapper,
           indexIO,
-          indexMerger
+          indexMerger,
+          rowIngestionMeters,
+          parseExceptionHandler
```
[the remark this answers is not in the document]
Comment (jihoonson) [reply server/src/main/java/org/apache/druid/segment/realtime/appenderator/PeonAppenderatorsManager.java:142]: They must not be null. I added null checks in `AppenderatorImpl`.

File: processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java [hunk 14]
```diff
@@ -186,7 +185,7 @@ protected AddToFactsResult addToFacts(
       } else {
         // We lost a race
         aggs = concurrentGet(prev);
-        parseExceptionMessages = doAggregate(metrics, aggs, rowContainer, row);
+        doAggregate(metrics, aggs, rowContainer, row, parseExceptionMessages);
```
[the remark this answers is not in the document]
Comment (jihoonson) [reply processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java:189]: Nice finding. Clearing parseExceptionMessages now.

File: processing/src/main/java/org/apache/druid/segment/incremental/OffheapIncrementalIndex.java (hunk 10 above)
Comment (jihoonson) [reply processing/src/main/java/org/apache/druid/segment/incremental/OffheapIncrementalIndex.java:236]: This is intentional to make it same with `OnheapIncrementalIndex`.

File: processing/src/main/java/org/apache/druid/segment/incremental/NoopRowIngestionMeters.java (hunk 11 above)
Comment (jihoonson) [reply processing/src/main/java/org/apache/druid/segment/incremental/NoopRowIngestionMeters.java:30]: This is used only in `RealtimeIndexTask` which is used by `Tranquility`. It can be a singleton, but doesn't seem matter.

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/BatchAppenderators.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (abhishekagarwal87) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/BatchAppenderators.java:70]: fwiw, I think current implementation is better one.

File: processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (suneet-s) [reply processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java:189]: why not just return a List<String> like before? That seems less risky than assuming the list is mutated correctly. In general, I'm not a big fan of returning values from a function using parameters because of cases like this.

Either way, this comment shouldn't be considered a blocker, just my 2 cents

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/StreamChunkParser.java [hunk 15]
```diff
@@ -54,12 +64,16 @@
       @Nullable InputFormat inputFormat,
       InputRowSchema inputRowSchema,
       TransformSpec transformSpec,
-      File indexingTmpDir
+      File indexingTmpDir,
+      Predicate<InputRow> rowFilter,
```
Comment (suneet-s) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/StreamChunkParser.java:68]: I don't think we should use a `Predicate<InputRow>` here and in other similar places. Throwing a `ParseException` is expected behavior, so I think callers should know that they need to handle that exception or throw it themselves.

It's not obvious to me how the caller in `FilteringCloseableInputRowIterator` should know they should handle a parse exception https://github.com/apache/druid/pull/10336/files#diff-b7f61f3d28afdd25bedf2efa607fdf88R67

How can I check that the ParseException is handled at the correct level everywhere? I realize this is challenging because ParseException is a RuntimeException.

**Review by suneet-s — COMMENTED** [review]
In general, a common question I've asked myself while reviewing this patch is are there any places where we now parse more rows vs failing earlier in the process by directly throwing an exception. It was hard for me to discern that reading this patch. Do you have any suggestions on how to look for cases where an exception is being swallowed instead of being thrown immediately?

File: processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (jihoonson) [reply processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java:189]: I'm not a big fan of that either, but it seems strange to me to return exception messages as a result of aggregation. I'm actually not sure if we can return early without the second aggregation when the first aggregation throws parseExceptions. I don't want to touch that part yet at least in this PR.

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/StreamChunkParser.java (hunk 15 above)
Comment (jihoonson) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/StreamChunkParser.java:68]: > I don't think we should use a Predicate<InputRow> here and in other similar places. Throwing a ParseException is expected behavior, so I think callers should know that they need to handle that exception or throw it themselves.
>
> It's not obvious to me how the caller in FilteringCloseableInputRowIterator should know they should handle a parse exception https://github.com/apache/druid/pull/10336/files#diff-b7f61f3d28afdd25bedf2efa607fdf88R67

Hmm, I'm not sure I understand your concern. Can you elaborate more?. `rowFilter` is a simple predicate to filter out unnecessary rows and does nothing with parseExceptions. Instead, it just passes the ParseException to the caller if it is thrown in `test()`. The intention here is that the callers of `FilteringCloseableInputRowIterator` can iterate rows without worrying about validation of rows or parseExceptions.

**Comment by jihoonson** [comment]
> In general, a common question I've asked myself while reviewing this patch is are there any places where we now parse more rows vs failing earlier in the process by directly throwing an exception. It was hard for me to discern that reading this patch. Do you have any suggestions on how to look for cases where an exception is being swallowed instead of being thrown immediately?

I mainly touched Kafka/Kinesis tasks and IndexTask and we have some tests for them which verify the metrics in `RowIngestionMeters`. Check out `IndexTaskTest`, `KafkaIndexTaskTest`, and `KinesisIndexTaskTest`. So I believe there is no bug in at least what are covered by those tests. 

One of the goals in this PR is clarifying where `ParseException` can be thrown. I first tried to change `ParseException` to be a regular `Exception` instead of `RuntimeException`, but it requires to modify all method signatures used on the query side even though `ParseException` will not be thrown from them. Since this doesn't look nice, I decided to scope down the places where can potentially throw `ParseException`s. After this PR, the only possible places are `InputEntityReader` (and all classes it uses) and `IncrementalIndex.addToFacts()` and any `ParseException` thrown in other places is a bug.

File: server/src/main/java/org/apache/druid/segment/realtime/appenderator/AppenderatorImpl.java [hunk 16]
```diff
@@ -267,6 +275,12 @@ public AppenderatorAddResult add(
       throw new SegmentNotWritableException("Attempt to add row to swapped-out sink for segment[%s].", identifier);
     }
 
+    if (addResult.isRowAdded()) {
+      rowIngestionMeters.incrementProcessed();
+    } else if (addResult.hasParseException()) {
+      parseExceptionHandler.handle(addResult.getParseException());
+    }
```
Comment (suneet-s) [inline server/src/main/java/org/apache/druid/segment/realtime/appenderator/AppenderatorImpl.java:282]: Test coverage bot is complaining about a lack of coverage for this condition. I think we should add tests for these branches. Right now `AppenderatorTest` only tests `addResult.isRowAdded`

Covering all 4 branches should make the code coverage bot happy too.

**Review by suneet-s — APPROVED** [review]
Code LGTM

I think we should make the code coverage bot happy in this case since the branches it's complaining about is in one of the key classes for this change.

```
|  71% (5/7)     |  25% (1/4)     | 100% (4/4)     | org/apache/druid/segment/realtime/appenderator/AppenderatorImpl.java
```

**Comment by suneet-s** [comment]
> ```
> 279 L                   rowIngestionMeters.incrementProcessed();
> 280 F | L | B(0/2)    } else if (addResult.hasParseException()) {
> 281 L                   parseExceptionHandler.handle(addResult.getParseException());
> 282 F                 }
> ```
> 
> The test coverage bot is failing because of the missing tests for these branches, but they are being covered by `IndexTaskTest` which is in `indexing-service`.

Oops sorry about my previous comment. How did you get code coverage to show you that these branches were covered? I tried running this locally, but it doesn't show me code coverage. This is what I see in the `indexing modules` tests

```
Diff coverage statistics:
------------------------------------------------------------------------------
|     lines      |    branches    |   functions    |   path
------------------------------------------------------------------------------
| 100% (1/1)     | 100% (0/0)     | 100% (1/1)     | org/apache/druid/indexing/common/stats/DropwizardRowIngestionMeters.java
| 100% (6/6)     | 100% (0/0)     | 100% (7/7)     | org/apache/druid/indexing/common/task/batch/parallel/PartialSegmentGenerateTask.java
| 100% (18/18)   | 100% (4/4)     | 100% (16/16)   | org/apache/druid/indexing/common/task/batch/parallel/PartialDimensionDistributionTask.java
|  80% (25/31)   |  70% (7/10)    |  93% (27/29)   | org/apache/druid/indexing/common/task/batch/parallel/SinglePhaseSubTask.java
|   0% (0/45)    |   0% (0/16)    |  25% (2/8)     | org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java
|   0% (0/1)     | 100% (0/0)     | 100% (1/1)     | org/apache/druid/indexing/seekablestream/SeekableStreamIndexTask.java
|  88% (15/17)   |  66% (4/6)     |  48% (12/25)   | org/apache/druid/indexing/seekablestream/StreamChunkParser.java
| 100% (14/14)   | 100% (2/2)     |  85% (24/28)   | org/apache/druid/indexing/common/task/AbstractBatchIndexTask.java
|  89% (17/19)   |  75% (6/8)     |  89% (33/37)   | org/apache/druid/indexing/common/task/InputSourceProcessor.java
| 100% (2/2)     | 100% (0/0)     | 100% (11/11)   | org/apache/druid/indexing/common/task/BatchAppenderators.java
|  90% (36/40)   |  85% (12/14)   |  83% (10/12)   | org/apache/druid/indexing/common/task/IndexTask.java
|  95% (21/22)   |  90% (9/10)    |  94% (51/54)   | org/apache/druid/indexing/common/task/FilteringCloseableInputRowIterator.java
|  80% (8/10)    |  50% (1/2)     |  80% (8/10)    | org/apache/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTask.java
| 100% (4/4)     | 100% (0/0)     |   0% (0/1)     | org/apache/druid/indexing/overlord/sampler/InputSourceSampler.java
------------------------------------------------------------------------------

```

**Comment by jihoonson** [comment]
Ah I didn't check the code coverage. I manually checked them instead. Well, not only in `IndexTaskTest`, but there are also many tests in `KafkaIndexTaskTest` and `KinesisIndexTaskTest` which verify `processed` and `unparseable` metrics. All those tests would fail if these lines didn't work properly.

File: processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (liran-funaro) [reply processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java:189]: Currently (before and after this PR), a row that contains partially corrupted/incorrect data is aggregated anyway (partially).
This may result in wrong statistics inference by the user due to inconsistencies between the number of rows each aggregator accumulated.
IMO, the decision of allowing this scenario should be of the user. 
But since currently, there is no way to roll-back aggregation, I don't see how this can be easily implemented.
Maybe adding parse as a preliminary step to aggregation, but this is an entirely different subject.
"return early without the second aggregation" would only aggravate this scenario, so I don't think it is the way to go either.

**Review by liran-funaro — COMMENTED** [review]
I'm just adding a couple of minor suggestions.

File: processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (jihoonson) [reply processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java:189]: Good point. I'm not sure if that behaviour is intended or not, but seems strange to me. I think all parse operations should be done in `InputSourceReader.read()` as a preliminary step so that we don't have to worry about ParseExceptions in IncrementalIndex. But yeah, this is a different subject.

File: processing/src/test/java/org/apache/druid/segment/incremental/OnheapIncrementalIndexBenchmark.java [hunk 17]
```diff
@@ -227,9 +224,7 @@ protected AddToFactsResult addToFacts(
           }
           catch (ParseException e) {
             // "aggregate" can throw ParseExceptions if a selector expects something but gets something else.
-            if (getReportParseExceptions()) {
-              throw e;
-            }
+            throw e;
```
[the remark this answers is not in the document]
Comment (jihoonson) [reply processing/src/test/java/org/apache/druid/segment/incremental/OnheapIncrementalIndexBenchmark.java]: Good catch. Fixed.

**Comment by jihoonson** [comment]
It turns out that I can easily add a new test for missing branches, so I added one. Hopefully the bot is happy now.

**Review by suneet-s — APPROVED** [review]
LGTM after CI. thanks @jihoonson !

**Comment by jihoonson** [comment]
@suneet-s thanks for the review! @abhishekagarwal87 @liran-funaro do you have more comments?

**Comment by abhishekagarwal87** [comment]
@jihoonson - Honestly, I haven't done a full review but +1 based on @suneet-s's review/approval.

**Comment by liran-funaro** [comment]
@jihoonson LGTM.

**Comment by jihoonson** [comment]
@abhishekagarwal87 @liran-funaro thanks for the review.
